### DISCUSSION TITLE
Understanding Layers

### DISCUSSION THREAD

**Discussion by cuppajoeman** [jrouwe/JoltPhysics#851] [Q&A] [answered]
Hi there,

I'm currently working through the hello world example and trying to understand layers in the physics system. I've read the jolt doxygen docs on the layers many times but it's not going through my head. I'm mainly stuck on the broad phase layer conceptually.

When I say layer here I'll be talking about the broadphase layer unless specified otherwise.

I'm trying think of a more concrete situation that can help me, so far this is what my understanding looks like:

> In Jolt we try and filter out as many possible collisions as possible. The process starts by using broadphase layers which are a way of allowing the programmer to say what collections of objects can possibly collide with eachother.

> An example of this could be we are working on an fps game and we have a stair set, one way of handling this could be to model the stairs using a wedge, and then render stairs in place of the wedge so that the character seems to move up the stairs smoothly.

> This is fine, but suppose the player shoots their gun and a shell comes out, instead of landing on the stairs and either falling down consecutive stairs or stopping, the shell would simply roll down the wedge and through the stairs which would would render as a visual glitch to the player. 

> This is where layers can come in for the rescue, instead we could have a navigation layer, and a detail layer, where given any player, they can only collide with things in the navigation layer, and then for bullet shells, they collide with the detail layer. Concretely players, along with simplified geometry representing the map are in the navigation layer, and bullet shells and detailed geometry of the map are in the detail layer.

At this point I get a little confused because I also need bullet shells to bounce off other players, this would mean that players also have to exist within the detail layer, or bullet shells have to exist in the navigation layer so that bullets and players know about eachother and can do collisions, otherwise the broaphase would never detect a collision between them and they would pass through eachother. I was also thinking another option could be to have another layer specifically for player-bullet shell collisions, but read that creating many layers is expensive so that probably isn't the best approach.

So instead suppose that we put our player in the detail layer, then I don't think this wouldn't be the best either because then the broadphase quadtree would check if the player is colliding with the detailed stairs again, and then that would defeat the purpose of separating them into layers? If we put the bullet into the navigation tree, then it would now check for collisions with the low detail ramp, which we also didn't want.

I think the object phase could handle this by just saying that complex map geometry object layer cannot collide with anything in the player object layer, but this wouldn't stop the quadtree collision checks done in the previous paragraph?

I was hoping someone could take my example and describe how they would set up all the layers and filters required for `physics_system.init`, thanks!

**Comment by jrouwe** [comment]
> At this point I get a little confused because I also need bullet shells to bounce off other players, this would mean that players also have to exist within the detail layer

This is incorrect. The player could exist in any broad phase layer. An object layer can collide with multiple broad phase layers and will simply iterate over all of the matching broad phase trees to find collisions.

I think you should not overthink the broad phase layers, the most important part is that there should be at least 1 layer for stuff that moves frequently and 1 layer for stuff that moves infrequently (assuming there's much more of the latter than the former). Everything beyond that is fine tuning.

**Reply by cuppajoeman** [reply]
Hi Jorrit, thanks for the response. 

I read your reply and just want to make sure I have this straight:

1. `BroadPhaseLayerInterface` represents a one to one mapping between object layers and broadphase layers.

2. `GetBroadPhaseLayer` is a method that gets the single broad phase layer an object layer resides/exist in.  The docs for this method say:
  > Convert an object layer to the corresponding broadphase layer. 

  So I'm not sure if my understanding is 100%

3. I'm having a hard time conceptualizing what's being mentioned in the docs: 

> These determine which object layers ([ObjectLayer](https://jrouwe.github.io/JoltPhysics/_object_layer_8h.html#a960eb396e2803ddad87bf25e0dcb4470)) collide with which broad phase/object layer 

The language in the docs make make it seem like a collection of objects (every object that reside in the same object layer) can collide with a broadphase layer, which is a quadtree with elements in it. I'm not really sure how that would work unless some bounding volume hierarchy is being used on all the objects in the object layer, and then we're checking if that structure collides with the quadtree? Could you explain what this language implies?

4. In the following reply:

> The player could exist in any broad phase layer. An object layer can collide with multiple broad phase layers and will simply iterate over all of the matching broad phase trees to find collisions.

Supposing we have  moving and non-moving broadphase layers, and then we have a player which is inside of the moving broadphase layer, then when we iterate over the moving broadphase layer (which is an AABB quadtree) I get that we can check for collisions because the player will exist inside of that quadtree, but since it won't exist inside of the non-moving quadtree, how can we check for collisions there?

Appreciate it !

**Reply by jrouwe** [reply]
I tried to answer your questions by expanding the [Collision Detection](https://jrouwe.github.io/JoltPhysics/#collision-detection) documentation and in particular the [Collision Filtering](https://jrouwe.github.io/JoltPhysics/#collision-filtering) section. Can you check if it makes sense now?

**Reply by cuppajoeman** [reply]
I read the new section a few times over, but honestly I still don't get how the system of layers works. The main confusion comes from reading this block of text: 

![image](https://github.com/jrouwe/JoltPhysics/assets/19198104/2c57ef1f-0876-4fd7-8bf9-30737b9dc834)

I think the reason it's so hard to understand is because it introduces many different definitions and data structures at all once, while also trying to motivate why we need layers simultaneously. I also find it hard to understand because it talks about collisions between layers and objects, and I don't know what that means.

I think the first paragraph before this section is a good starting point because it clearly states one thing well. We should make sure that the second paragraph does the same. 

Below this text you will find how I might structure the docs to be least confusing and explain clearly why this system exists along with clear definitions of what we're talking about.

**NOTE: the text below comes from my understanding of the jolt layers/filters system, and may be incorrect**
---

### Motivation

Give an Example of how the system would work with no layers then show why having layers is beneficial. During the motivation we explain how collisions are conducted regularly, by querying the AABBQuadTree, with a given shape (or however that works).



### Definitions and Properties 

Here we define what the following terms are.

#### entities
* object: 
  * definition: A body (this begs the question, why don't we skip "object" all together and call the next thing a BodyLayer ?)
  * property: Every object must exist in exactly one object layer
* object layer
  *  definition: a collection of objects that all have the same ObjectLayer attribute
  * property: Every object layer must exist in exactly one broadphase layer
* BroadPhaseLayer
  * definition: A collection of object layers along with an AABBQuadtree, such that every object in each object layers is stored in the tree. This means that if object layer A and object layer B are in the same broadphase layer, then every object in A and every object in B reside in this tree.
* BroadPhaseQuadTree
  * definition: A collection of broadphase layers
* BroadPhaseLayerInterface
  * definition: For each object layer, this object holds which broadphase layer it exists in through [BroadPhaseLayerInterface::GetBroadPhaseLayer](https://jrouwe.github.io/JoltPhysics/class_broad_phase_layer_interface.html#a5007c9b9beab5c16af36e89b48cd06cb), this respects the object layers property.
* ObjectVsBroadPhaseLayerFilter
  * definition: defines whether or not an object from object layer L, can collide with objects in a specific BroadPhaseLayer 
* ObjectLayerPairFilter
  * definition: defines whether or not an object from object layer L can collide with an object from object layer K

#### gotcha's
* Just because an object can exist in exactly one broadphase layer, we can still check if a given object collides with other objects located in other broadphase layers.

#### collisions
* potential object-object collisions (because this is broadphase)
  * definition: when the AABB boxes of the two objects intersect
  * property: when two objects AABB boxes collide it's likely that the objects collide, but if their boxes don't intersect it's impossible for them to collide
* potential object - object layer collisions:
  * definition: when the object potentially collides with at least one of the objects stored in the layer.
  * How we do it: Iterate through each element in the layer and check if they collide (how?)
* potential object layer - object layer collisions:
  * definition: when an object layers potentially collides with the other object layers bounding box (this seems wrong, what is an object layer bounding box)
  * How we do it: Get the bounding box for the two object layers and check if they collide (not sure about that)
* potential object layer - broadphase layer collisions:
  * definition: We say that an object layer and a broadphase layer are potentially colliding when the object layer potentially collides with at least one of the object layers stored in the broadphase layer.

### Explain the Collision Pipeline clearly
* We start by iterating over every active body in the world.
  * For a given body we figure out what object layer it exists in, call this layer L
  * For each broadphase layer B in the BroadPhaseQuadTree, we check if a body with object layer L can collide with broadphase layer B via ObjectVsBroadPhaseLayerFilter, if not go back to step one
  * Otherwise, then we go check for collisions with this object in this broadphase layers AABB quadtree
    * For each potential object-object collision detected in this tree, we check if this collision can occur via ObjectLayerPairFilter, if not skip it, otherwise it's marked by the broadphase for the next narrowphase step.

(I skipped over other filters such as body, shape and group filters here, because I don't know too much about them now)

**Reply by jrouwe** [reply, marked as answer]
So far there have been very few questions about the layer system, most people just seem to get it. I don't have the time to keep refining this bit of documentation and I also don't want to spend my time going over changes you propose to this document (it costs me a lot of time just reading the wall of text above).

What I suggest is that you just stick to having a MOVING and NON_MOVING broad phase layer and map your object layers to one of these 2 layers depending on if they're movable or not. Use either the ObjectLayerPairFilterTable or ObjectLayerPairFilterMask (and friends) implementations so you don't have to care about the details.
